browser/brave_ads/tooltips/ads_tooltips_controller.cc

```diff
@@ -0,0 +1,57 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_ads/tooltips/ads_tooltips_controller.h"
+
+#include <memory>
+#include <utility>
+
+#include "brave/browser/ui/brave_tooltips/brave_tooltip_popup.h"
+
+namespace brave_ads {
+
+AdsTooltipsController::AdsTooltipsController(Profile* profile)
+    : profile_(profile) {
+  DCHECK(profile_);
+}
+
+AdsTooltipsController::~AdsTooltipsController() = default;
+
+void AdsTooltipsController::ShowTooltip(
+    std::unique_ptr<brave_tooltips::BraveTooltip> tooltip) {
+  DCHECK(tooltip);
+
+  // If there's no delegate, set one so that clicks go back to the appropriate
+  // handler
+  tooltip->set_delegate(AsWeakPtr());
+
+  const std::string tooltip_id = tooltip->id();
+  DCHECK(!tooltip_popups_[tooltip_id]);
+  tooltip_popups_[tooltip_id] =
+      new brave_tooltips::BraveTooltipPopup(profile_, std::move(tooltip));
```

Comment:
Who's the owner of these `BraveTooltipPopup`s?

Reply:
They live in a map in this controller and this controller exists for the lifetime of the ads service.

Reply:
Then we're leaking those in `OnTooltipWidgetDestroyed()`, aren't we?

Reply:
They're erased and destroyed in `OnTooltipWidgetDestroyed`, unless I'm just misunderstanding.

Reply:
You store raw pointers in `tooltip_popups_`.

Reply:
Ahh I see, now I have to remember why I did that lol. Thanks for point it out!

Reply:
cc: @tmancey: As this was based on `AdNotificationPopup`, but it sounds like maybe this is just how widget ownership is handled?

Reply:
@szilardszaloki the view is owned by the widget and confirm this I added `VLOG` to the constructor and destructor for `AdNotificationPopup`

Reply:
Just curious - I can see where `AdNotificationPopup`s are getting allocated:
https://github.com/brave/brave-core/blob/fb3cee825307a0d9d03692bd11832e00c08cfadb/browser/ui/brave_ads/ad_notification_popup.cc#L122-L138 via https://github.com/brave/brave-core/blob/9f89e03a91a54fcd06ec8a3e657b88be1f4b7a49/browser/ui/brave_ads/ad_notification_popup.cc#L77-L81 but I can't seem to find where they're getting deallocated.

Reply:
@szilardszaloki see `widget->InitWidget(this, CalculateBounds());` in `CreateWidgetView`

Reply:
@tmancey Gotcha! And the same applies to `BraveTooltipPopup::CreateWidgetView()` in @emerick's case, I suppose.

Reply:
If so, then @emerick, would you mind leaving a `// NOT OWNED` comment for `tooltip_popups_` (`std::map<std::string, brave_tooltips::BraveTooltipPopup* /*NOT OWNED*/> tooltip_popups_;`) and maybe another one for `tooltip_popups_.erase(tooltip_id);`, saying something like "the pointed-to `BraveTooltipPopup`s get deallocated by their containing `Widget`s"? I think it would be of great help in understanding the details here.

Reply:
Good idea, done.

Reply:
@szilardszaloki that is correct! Thanks

Reply:
@emerick would it be ok to also add the same comments to the ad notifications in the same PR?

Reply:
Yes, good idea - added.